ethereum/dataproviders/src/test/java/tech/pegasys/teku/dataproviders/lookup/BlockProviderTest.java

```diff
@@ -132,4 +141,50 @@ void fromDynamicMap() {
     assertThat(provider.getBlock(blockB.getRoot()))
         .isCompletedWithValue(Optional.of(blockB.getBlock()));
   }
+
+  @Test
+  void fromMapWithLock() throws ExecutionException, InterruptedException {
+    chainBuilder.generateGenesis();
+    final Map<Bytes32, SignedBeaconBlock> blocks =
+        chainBuilder
+            .streamBlocksAndStates(0)
+            .collect(Collectors.toMap(SignedBlockAndState::getRoot, SignedBlockAndState::getBlock));
+    // 1 block is ok for this test
+    assertThat(blocks.size()).isEqualTo(1);
+    final SignedBeaconBlock block = blocks.values().stream().findFirst().orElseThrow();
+
+    final ReadWriteLock lock = new ReentrantReadWriteLock();
+    final Lock readLock = lock.readLock();
+    final BlockProvider provider = BlockProvider.fromMapWithLock(blocks, readLock);
+
+    // Read block from provider
+    assertThat(provider.getBlock(block.getRoot()).get()).contains(block);
+
+    // Try the same when locked
+    final AsyncRunner asyncRunner = DelayedExecutorAsyncRunner.create();
+    final CountDownLatch isNotBlocked = new CountDownLatch(1);
+    final AtomicBoolean notBlockedOnFuture = new AtomicBoolean(false);
+    // Write lock in main thread
+    lock.writeLock().lock();
+    try {
+      asyncRunner
+          .runAsync(
+              () -> {
+                try {
+                  final SafeFuture<Optional<SignedBeaconBlock>> blockFuture =
+                      provider.getBlock(block.getRoot());
+                  notBlockedOnFuture.set(true);
+                  // Attempt to read in other thread - will stuck here
+                  assertThat(blockFuture.get()).contains(block);
+                } finally {
+                  isNotBlocked.countDown();
+                }
+              })
+          .ifExceptionGetsHereRaiseABug();
+      assertThat(isNotBlocked.await(1, TimeUnit.SECONDS)).isFalse();
```

Comment:
even something small like 100ms should be plenty here?

Reply:
Sure